### ISSUE TITLE
Keyboard key press and key release events

### ISSUE THREAD

**Issue by yuvallanger** [hajimehoshi/ebiten#648]
Add a way to detect a key press event and a key release event.

In my particular problem I am trying to toggle the display of debug information on screen. I have a global `var debug bool`. It is being toggled when I press D:

```
if ebiten.IsKeyPressed(ebiten.KeyD) {
    debug = !debug
}
```

the debug information is displayed:

```
if debug == true {
    ebitenutil.DebugPrint(…)
}
```

Problem here is that at each iteration of the main loop `debug` is toggled at about 60 Hz… A key press and release events would help to solve that.

**Comment by atolVerderben** [comment]
I believe this is covered in the [typewriter example](https://hajimehoshi.github.io/ebiten/examples/typewriter.html)

Relevant code:

> // repeatingKeyPressed return true when key is pressed considering the repeat state.
func repeatingKeyPressed(key ebiten.Key) bool {
	const (
		delay    = 30
		interval = 3
	)
	d := inpututil.KeyPressDuration(key)
	if d == 1 {
		return true
	}
	if d >= delay && (d-delay)%interval == 0 {
		return true
	}
	return false
}

Specifically I believe you could use inpututil.KeyPressDuration(ebiten.KeyD) for exactly your needs.

**Comment by yuvallanger** [comment]
Many thanks!

**Comment by yuvallanger** [comment]
Oh boy. I didn't realize that there's even more under the child directories.

I've just checked and there are two functions that seem to provide my wanted functionality:

https://godoc.org/github.com/hajimehoshi/ebiten/inpututil#IsKeyJustPressed
https://godoc.org/github.com/hajimehoshi/ebiten/inpututil#IsKeyJustReleased

I have checked `inpututil.IsKeyJustPressed` and sometimes when I release the pressed button, the event is fired again. That is once when in press and once when I release. If the time between press and release is small, the event is fired only when I press.

**Comment by hajimehoshi** [comment]
Oh, you mean that the even for release is not fired when you press a key too quickly? That might be a bug. Which environment are you using? Thanks!

**Comment by yuvallanger** [comment]
```
$ gopherjs version
GopherJS 1.10-4
$ go version
go version go1.10.3 linux/amd64
```

Git's HEAD at my `$GOPATH/src/github.com/hajimehoshi/ebiten` points at the
commit `9047e54ae6541ceb6a5415f0e5ecc563dbaa04dc`.

The problem only appears in the go version, not the gopherjs web version.

Please ask me for more information as needed.

**Comment by hajimehoshi** [comment]
```
package main

import (
        "github.com/hajimehoshi/ebiten"
        "github.com/hajimehoshi/ebiten/inpututil"
)

func update(screen *ebiten.Image) error {
        if inpututil.IsKeyJustPressed(ebiten.KeyA) {
                println("A is just pressed")
        }
        if inpututil.IsKeyJustReleased(ebiten.KeyA) {
                println("A is just released")
        }
        return nil
}

func main() {
        if err := ebiten.Run(update, 320, 240, 2, "foo"); err != nil {
                panic(err)
        }
}
```

I couldn't reproduce this on my MacBook Pro.

**Comment by yuvallanger** [comment]
I've tried reproducing using your code and it did reproduce:

```
$ go run reproduce.go 
# github.com/go-gl/glfw/v3.2/glfw
In file included from /home/yuval/gopath/src/github.com/go-gl/glfw/v3.2/glfw/c_glfw_linbsd.go:24:0:
/home/yuval/gopath/src/github.com/go-gl/glfw/v3.2/glfw/glfw/src/linux_joystick.c: In function ‘_glfwInitJoysticksLinux’:
/home/yuval/gopath/src/github.com/go-gl/glfw/v3.2/glfw/glfw/src/linux_joystick.c:224:42: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 9 [-Wformat-truncation=]
             snprintf(path, sizeof(path), "%s/%s", dirname, entry->d_name);
                                          ^~~~~~~
In file included from /usr/include/stdio.h:862:0,
                 from /usr/include/X11/Xcursor/Xcursor.h:26,
                 from /home/yuval/gopath/src/github.com/go-gl/glfw/v3.2/glfw/glfw/src/x11_platform.h:39,
                 from /home/yuval/gopath/src/github.com/go-gl/glfw/v3.2/glfw/glfw/src/internal.h:169,
                 from /home/yuval/gopath/src/github.com/go-gl/glfw/v3.2/glfw/glfw/src/x11_init.c:28,
                 from /home/yuval/gopath/src/github.com/go-gl/glfw/v3.2/glfw/c_glfw_linbsd.go:19:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:64:10: note: ‘__builtin___snprintf_chk’ output between 12 and 267 bytes into a destination of size 20
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

here I press `a` and quickly release it

```
A is just pressed
A is just released
```

here I press it and release only after a while:

```
A is just pressed
A is just released
A is just pressed
A is just released
```

```
$ uname -a
Linux yankel-Inspiron-5567 4.15.0-24-generic #26-Ubuntu SMP Wed Jun 13 08:44:47 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04 LTS
Release:        18.04
Codename:       bionic
```

**Comment by hajimehoshi** [comment]
So the problem is that only one time pressing and releasing a key print out four messages like

```
A is just pressed
A is just released
A is just pressed
A is just released
```

?

**Comment by yuvallanger** [comment]
That's right.

**Comment by hajimehoshi** [comment]
Thanks.

Hmm, this might be a hardware or a driver problem on your keyboard. If you pressing A, is "A" shown only once? http://en.key-test.ru/ (as browser implementation might be different from GLFW, this might hide the problem though).

**Comment by yuvallanger** [comment]
The problem does not appear in the browser.

**Comment by hajimehoshi** [comment]
Ahh, the problem itself doesn't happen on GopherJS version as you mentioned, so testing the website doesn't make sense. Sorry for bothering you.

I appreciate your reporting. I'll take a look this on my Linux machine later.

**Comment by Noofbiz** [comment]
This sounds like this [issue in glfw](https://github.com/glfw/glfw/issues/747). It's been fixed for 3.3 but 3.3 is still in development :(

**Comment by SolarLune** [comment]
Dang, I think I'm running into this, too - `ebiten.IsKeyPressed()` is returning true even after I release the key. Is there some way around this, I wonder?

**Comment by hajimehoshi** [comment]
I also wonder, but have not come up with it.

**Comment by hajimehoshi** [comment]
I tested on my Linux laptop (DELL Inspiron 3162), but could not reproduced. Hmm

**Comment by hajimehoshi** [comment]
Ooh, ok, I could reproduce this by holding the key for long time. I'll think an ad-hoc way to fix this, or forking glfw.

**Comment by kibertoad** [comment]
@hajimehoshi Do you plan to contribute patch upstream? Probably maintaining fork indefinitely is not a very sustainable solution :)

**Comment by hajimehoshi** [comment]
I didn't think so, but it depends on go-gl/glfw's policy. As keeping go-gl/glfw should wrap the original GLFW as it is, I think I should NOT apply the change, but I should have asked the authors before I determined.

@shurcooL Hi, we would like to fix the bug in GLFW 3.2 by applying a patch to GLFW before having GLFW 3.3 which we are not sure when it is released, but I am not sure I can do that since go-gl/glfw might have to be just a simple wrapper to wrap the original GLFW. What do you think?

**Comment by hajimehoshi** [comment]
Created a branch to use hajimehoshi's GLFW https://github.com/hajimehoshi/ebiten/tree/issue-648-glfw-fork

**Comment by hajimehoshi** [comment]
@shurcooL I posted https://github.com/go-gl/gl/issues/111, so let's discuss there. Sorry for the confusion.

**Comment by hajimehoshi** [comment]
The fix was merged into go-gl/glfw.

@kibertoad Hi, thank you for pointing it out!

**Comment by kibertoad** [comment]
Woohoo!

**Comment by hajimehoshi** [comment]
Note: if you don't use Go modules, the below command should work:

```
go get -u github.com/hajimehoshi/ebiten/...
go get -u github.com/go-gl/glfw/v3.2/...
# go get -u github.com/go-gl/glfw/... failed since v3.0 cannot be compiled :-(
```

If you have a trouble, try clearing the cache:


```
go clean -i -cache github.com/go-gl/glfw/v3.2/...
```
